src/Interpreters/TransactionLog.cpp

```diff
@@ -0,0 +1,364 @@
+                zookeeper = new_zookeeper;
+            }
+
+            loadNewEntries();
+        }
+        catch (const Coordination::Exception & e)
+        {
+            LOG_ERROR(log, getCurrentExceptionMessage(true));
+            /// TODO better backoff
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            if (Coordination::isHardwareError(e.code))
+            {
+                std::lock_guard lock{mutex};
+                zookeeper.reset();
+            }
+            log_updated_event->set();
+        }
+        catch (...)
+        {
+            LOG_ERROR(log, getCurrentExceptionMessage(true));
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            log_updated_event->set();
+        }
+    }
+}
+
+void TransactionLog::loadNewEntries()
+{
+    Strings entries_list = zookeeper->getChildren(zookeeper_path, nullptr, log_updated_event);
+    assert(!entries_list.empty());
+    std::sort(entries_list.begin(), entries_list.end());
+    auto it = std::upper_bound(entries_list.begin(), entries_list.end(), last_loaded_entry);
+    loadEntries(it, entries_list.end());
+    assert(last_loaded_entry == entries_list.back());
+    assert(latest_snapshot == parseCSN(last_loaded_entry));
+    latest_snapshot.notify_all();
+}
+
+
+Snapshot TransactionLog::getLatestSnapshot() const
+{
+    return latest_snapshot.load();
+}
+
+MergeTreeTransactionPtr TransactionLog::beginTransaction()
+{
+    MergeTreeTransactionPtr txn;
+    {
+        std::lock_guard lock{running_list_mutex};
+        Snapshot snapshot = latest_snapshot.load();
+        LocalTID ltid = 1 + local_tid_counter.fetch_add(1);
+        txn = std::make_shared<MergeTreeTransaction>(snapshot, ltid, ServerUUID::get());
+        bool inserted = running_list.try_emplace(txn->tid.getHash(), txn).second;
+        if (!inserted)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "I's a bug: TID {} {} exists", txn->tid.getHash(), txn->tid);
+        txn->snapshot_in_use_it = snapshots_in_use.insert(snapshots_in_use.end(), snapshot);
+    }
+
+    LOG_TEST(log, "Beginning transaction {} ({})", txn->tid, txn->tid.getHash());
+    tryWriteEventToSystemLog(log, global_context, TransactionsInfoLogElement::BEGIN, txn->tid);
+
+    return txn;
+}
+
+CSN TransactionLog::commitTransaction(const MergeTreeTransactionPtr & txn)
+{
+    txn->beforeCommit();
+
+    CSN new_csn;
+    /// TODO Transactions: reset local_tid_counter
+    if (txn->isReadOnly())
+    {
+        LOG_TEST(log, "Closing readonly transaction {}", txn->tid);
+        new_csn = txn->snapshot;
+        tryWriteEventToSystemLog(log, global_context, TransactionsInfoLogElement::COMMIT, txn->tid, new_csn);
+    }
+    else
+    {
+        LOG_TEST(log, "Committing transaction {}{}", txn->tid, txn->dumpDescription());
+        /// TODO handle connection loss
+        /// TODO support batching
+        auto current_zookeeper = getZooKeeper();
```

Comment:
Yes, because it's unclear how to maintain two separate logs for replicated and non-replicated MergeTree tables (especially, when both engines can be used in single transaction). But it's possible to use local one-node Keeper.